### PR TITLE
fix(cache): key proxy adapters by environmentName in preview mode

### PR DESCRIPTION
**Pull request by kwakayama**
## Problem

`ProxyFSAdapterManager` asserts that a reused adapter's identity — including `environmentName` — matches the request (`proxy-manager.ts:363,367`), but the non-production branch of `buildProxyManagerCacheKey` omitted `environmentName` from the key (`render.ts:140`).

Callers that resolve `environmentName` inconsistently (`null` vs `"preview"`) therefore collide on a single cache entry, and every reuse throws a fatal identity mismatch.

## Impact

Preview rendering is **fully down on staging** — 6/6 projects tested return `500 Cache path invariant violated`:

```
FATAL: Identity mismatch for cached adapter.
Reason: Cached environmentName does not match the request
cachedIdentity: { environmentName: null,      branch: "main" }
expected:       { environmentName: "preview", branch: "main" }
```

Server logs show **112 `null` vs 35 `"preview"`** resolutions in a 20-minute window, on the same request URL. It does not self-heal: the cache is in-memory, and both `veryfront-server` pods re-poisoned within minutes of their last restart.

## Fix

Include `environmentName` in the non-production key. The segment is omitted when unnamed, so existing keys stay byte-for-byte stable.

## Tests

Written red-green. The new tests fail on `main` with the exact production collision:

```
AssertionError: Expected actual: "proxy:example-project:preview:main"
                     not to be: "proxy:example-project:preview:main"
```

Three cases: `null` vs `"preview"` must differ, two distinct environment names must differ, and the unnamed key must stay stable.

`src/cache/keys.test.ts` passes 89/89. Note `src/cache/backend.test.ts` has 10 failures, but they are **pre-existing on clean `origin/main`** (verified 59 passed / 10 failed there) and unrelated to this change.

## Scope

`buildProxyManagerCacheKey` has one production caller, `proxy-manager.ts`.

## Follow-up

This stops the 500s but does not address *why* the same preview request resolves `environmentName` to `null` sometimes. That inconsistency is still worth tracing; with this change it costs an extra adapter rather than a fatal error.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved preview environment caching to keep separate environments from sharing cached data.
  * Preserved existing cache behavior for unnamed preview environments.
  * Improved adapter selection and initialization for named preview environments.
  * Ensured production requests consistently use production adapter settings.

* **Tests**
  * Added coverage confirming cache keys distinguish preview environments and handle omitted or null environment names consistently.
  * Added coverage for adapter reuse, environment normalization, and preview initialization.

### CODE REVIEW

**Comment by kwakayama** [comment]
## Automated review — round 1: 78/100, REQUEST_CHANGES

Posting the sub-agent review verdict and what was done about it.

| Sev | Location | Defect | Status |
|---|---|---|---|
| P1 | `hmr.handler.ts:238-252` | Warms the adapter without `environmentName` — the actual source of the `null` resolutions. Left as-is, it would warm a *different* adapter than renders use once the key includes `environmentName`, silently breaking poke reception | **Fixed** |
| P1 | `render.ts:128-137` / `proxy-manager.ts:357-365` | Production keys omit `branch` while the identity asserts on it — same collision class, reachable in production | **Fixed** |
| P2 | `render.ts:143` | `""` still collides with `null`: the key uses a truthiness guard, the identity used `?? null` | **Fixed** |
| P3 | `keys.test.ts:356-362` | Byte-stability test was tautological — compared the builder against itself | **Fixed** |
| P3 | `proxy-manager.test.ts` | No identity-mismatch coverage at the layer that actually 500'd | **Fixed** |
| P3 | `keys.test.ts` | Delimiter-injection property untested | **Fixed** |

Reviewer verified clean: encoder choice is injection-safe, and the backwards-compat claim holds.

### Notes on the two P1s

The HMR one is the important one. `resolvedEnvironment` was already destructured and used in the log line directly above the call — it just wasn't passed through. **This means the first commit alone would have traded a 500 for silently broken HMR**, so it belongs in the same PR.

For the production `branch` gap I took the narrower of the two suggested fixes: pin `effectiveBranch` to `null` in production rather than folding branch into the production key. Verified safe — production builds its content context from `releaseId`/`environmentName` only and never reads branch (`proxy-manager.ts:451-460`). This avoids invalidating production cache keys.

### Verification

Each fix was driven red-green. The HMR test failed with `undefined` vs `"preview"` before the one-liner. The empty-string and production-branch tests failed at the `getAdapter` layer before their fixes.

- `deno test` — 185 steps green across `keys.test.ts`, `proxy-manager.test.ts`, `hmr.handler.test.ts`
- `deno check`, `deno lint`, `deno fmt --check` — clean
- Round-1 CI was fully green (25 pass / 0 fail)

Re-review queued.

**Comment by kwakayama** [comment]
## Automated review — round 2: 64/100, REQUEST_CHANGES

The score went **down**, and correctly so: my round-1 fix for the HMR finding was wrong and regressed the common case.

### The blocker

`resolvedEnvironment` and `environmentName` are distinct fields (`src/types/server.ts:66,72`):

```ts
environmentName?: string;                            // the named environment
resolvedEnvironment?: "preview" | "production";      // the mode
```

I passed the **mode** where the **name** belongs. Because the guard above already narrows the mode to `"preview"`, the value was a constant. So HMR warmed `"preview"` while renders warm the actual environment name (e.g. `"Development"`) — different cache keys, and HMR's per-adapter `WebSocketManager` invalidating an adapter no render reads.

That is worse than `main` for any project without a named environment. Fixed in af94de7 by passing `ctx.environmentName ?? null`.

The round-1 test could not catch it: the fixture omitted `environmentName` and asserted the constant. Rewritten with a named environment (`"Development"`) distinct from the mode, so it fails when the two are confused — verified failing before the fix, passing after.

### Confirmed fixed from round 1

| Finding | Status |
|---|---|
| P1(b) production `branch` pinning | Confirmed — reviewer traced `createAdapter`, `getContextMismatchReason`, `matchesAdapterSelection`; none read branch in production |
| P2 empty-string normalization | Confirmed correct and whitespace-safe |
| P3 tautological byte-stability test | Confirmed |
| P3 `ProxyFSAdapterManager` identity coverage | Confirmed |
| P3 delimiter-injection test | Confirmed |

Also independently verified: the cache-key change keeps existing keys byte-identical.

### Known follow-up, not in this PR

`releaseId` in preview mode is the unfixed mirror of the branch bug — asserted in the identity, absent from the preview key, unnormalized. Rated LOW because upstream normalization masks it today, so it is not currently reachable. Worth closing separately rather than widening this PR.

### Verification

- 185 test steps green across `keys.test.ts`, `proxy-manager.test.ts`, `hmr.handler.test.ts`
- `deno check`, `deno lint`, `deno fmt` clean
- CI 25 pass / 0 fail. The earlier `tests (rsc browser e2e)` red was infrastructure — it died in *Install Chromium* with `Timed out waiting for apt/dpkg locks` before any test ran, and passed on rerun.

Round 3 review queued. Not merging until it clears 90.

**Comment by kwakayama** [comment]
## Automated review — round 3: 88/100, APPROVE

Approved, but below the 90 bar, so not merged. Both items the reviewer named for 90+ are now done in f219f5c.

### Fixed

**MEDIUM `proxy-manager.ts:126`** — `releaseId` was the last field asserted in the adapter identity but absent from the preview key: the same asymmetry this branch already fixed for `branch` and `environmentName`. Only upstream normalization in `multi-project-adapter.ts:81` kept it unreachable — nothing in `getAdapter` or the schema enforced that. Now pinned to `null` outside production.

**MEDIUM `proxy-manager.test.ts:175`** — "reuses a preview adapter without a fatal identity mismatch" passed unmodified against `origin/main` while its three siblings failed. It guarded nothing. Replaced with an interleaved unnamed → named → unnamed sequence. Verified it now fails with `render.ts` reverted to `main`:

```
FAILED | 0 passed (55 steps) | 1 failed (3 steps)
  ... reuses an unnamed preview adapter after a named one is created
```

**LOW `:213`, `:241`** — `assertExists` only proved "did not throw". Now `assertStrictEquals` on the adapter instances.

**LOW `hmr.handler.test.ts:562`** — stub set `isVeryfrontAdapter: true` where `wrapper.ts:115` calls it as a method. Now `() => true`. It passed before only because the guard is an `in`-check.

### Deliberately not changed

- **`?? null` at `hmr.handler.ts:254`** — redundant, but explicit at a call site where confusing two similarly-named fields already caused one regression on this branch.
- **`projectSlug` raw interpolation (`render.ts:137,146`)** and **schema `superRefine` asymmetry** — both pre-existing and rated not exploitable. Out of scope here.

### Identity audit — now closed

Every `ProxyAdapterIdentity` field is either in the key or normalized to match it:

| Field | Status |
|---|---|
| `projectSlug` | in key |
| `projectId` | in key via `authorityKey`, encoded |
| `credentialPrincipal` | in key, encoded, suffix position |
| `productionMode` | in key as the `mode` segment |
| `releaseId` | production: in key; preview: pinned `null` **(this commit)** |
| `environmentName` | in key when truthy, both modes, delimiter-escaped |
| `branch` | preview: in key; production: pinned `null` |

### Verification

- 186 test steps green across the three suites; each new test verified failing with its fix reverted
- `deno check`, `deno lint`, `deno fmt` clean
- CI on the prior commit: 25 pass / 0 fail

Round 4 queued.

**Comment by kwakayama** [comment]
## Automated review — round 4: 91/100, APPROVE

Clears the merge bar. Every prior fix was re-verified by reverting it individually, not just re-read.

### Both round-3 blockers closed

**`releaseId` pin** — the reviewer traced all seven downstream consumers (`createAdapter`, content-context build, `getContextMismatchReason`, `getIdentityMismatchReason`, `matchesAdapterSelection`, `buildDiagnosticCacheKey`/`getStats`, logging) and confirmed none reads `releaseId` on the preview path, so the pin is content-neutral.

Notably, this was **not** the unreachable edge case round 3 assumed: `src/server/handlers/request/rsc/index.test.ts:155` proves preview + non-null `releaseId` is a live reachable path that was previously producing a spurious `CACHE_INVARIANT_VIOLATION` plus forced eviction. Good thing it got fixed rather than deferred.

**Vacuous test** — the replacement fails with `render.ts` reverted to `origin/main`, and `assertStrictEquals(first, again)` now proves cache-slot identity rather than mere absence of a throw.

### Prior items — all re-verified by revert

| Item | Verified |
|---|---|
| `environmentName` keying | 3 of 4 keys tests fail on revert |
| empty-string normalization | test fails on revert |
| production `branch` pinning | test fails on revert |
| HMR named-environment fix | test fails on revert |
| 3 test-quality items | `assertExists`→`assertStrictEquals` at 4 sites; stub method matches `wrapper.ts:57,238`; vacuous test replaced |

### Remaining findings — all LOW

Addressed in 31cdf42: the normalization comment said "Both" while governing three normalizations and sat below the first one it described.

Not changed, with reasons:
- **Discarded preview `releaseId` is not logged.** Adding it would restore visibility of upstream misuse that previously threw loudly. Deliberately skipped to keep this PR minimal — worth doing, but it is new behaviour, not a fix.
- **`request-context.ts:84` pins `branch` but not `releaseId`.** Same asymmetry class, harmless now because of the downstream pin. Better as its own change than widened into this one.
- **`projectSlug` raw interpolation (`render.ts:137,146`)** — pre-existing, no constructible collision found.
- **Two encoders for `environmentName` across modes** (`render.ts:144` vs `source-identity.ts:43`) — pre-existing shape difference; unifying would change production keys.

### Process note

Rounds 1 and 2 used `git stash` to test reverts and swept up unrelated concurrent work from other sessions. Rounds 3 and 4 used `git checkout origin/main -- <file>` instead and created **zero** stashes. `stash@{0}` on this branch is that earlier leftover and holds another session's `run-stream.ts`/`run-stream.test.ts` work — **inspect it before dropping it.**

### Verification

- 186 test steps green; each new test verified failing with its fix reverted
- `deno check`, `deno lint`, `deno fmt` clean; working tree clean at `f219f5c`
- The two CI reds on this PR were both infrastructure: `Timed out waiting for apt/dpkg locks` in *Install Chromium*, before any test ran, in different jobs each time. Both passed on rerun.

Merging once CI is green on 31cdf42.

File: src/cache/keys/builders/render.ts [hunk 1]
```diff
@@ -138,7 +138,12 @@ export function buildProxyManagerCacheKey(
   }
 
   const source = encodeCacheSourceIdentity({ type: "branch", branch: branch ?? "main" });
-  return `${CacheKeyPrefix.PROXY}:${projectSlug}:${mode}:${source.qualifier}${authorityKey}`;
+  // ProxyFSAdapterManager asserts environmentName matches on reuse, so it must
+  // be part of the key. Omitted when unnamed to keep existing keys stable.
+  const environmentQualifier = environmentName
+    ? `:env:${encodeCacheKeyLiteralSegment(environmentName)}`
+    : "";
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/cache/keys/builders/render.ts:145]: Addressed — the collision is closed on the identity side rather than in the key.

`proxy-manager.ts:129` now normalizes with the same truthiness predicate the key builder uses:

```ts
const effectiveEnvironmentName = environmentName || null;
```

So `""` and `null` produce both the same key *and* the same identity — they can no longer diverge, which was the actual failure mode (the key omitted the segment while `?? null` kept `""` in the identity, so the reuse assertion threw).

The other call site, `buildDiagnosticCacheKey` at `proxy-manager.ts:52`, is fed already-normalized identity fields, so it cannot reintroduce the split.

Covered by `proxy-manager.test.ts` "treats an empty environment name as unnamed", which asserts `assertStrictEquals(empty, unnamed)` and fails when the normalization is reverted.

File: src/platform/adapters/fs/veryfront/proxy-manager.test.ts [hunk 2]
```diff
@@ -138,6 +138,122 @@ describe("ProxyFSAdapterManager", () => {
...
+          "my-project",
+          "test-token",
+          undefined,
+          false,
+          null,
+          null,
+          "main",
+        );
+
+        assertExists(unnamed);
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/platform/adapters/fs/veryfront/proxy-manager.test.ts]: Agreed, and already done in f219f5c — you were reviewing an earlier commit.

All four normalization tests now capture the first adapter and assert reuse with `assertStrictEquals` instead of `assertExists`:

- `:213` `assertStrictEquals(first, again)`
- `:241` `assertStrictEquals(empty, unnamed)`
- `:269` `assertStrictEquals(withRelease, withoutRelease)`
- `:297` `assertStrictEquals(withBranch, withoutBranch)`

Your point was exactly right: `assertExists` only proved the call did not throw, so it would have passed even if the second lookup built a fresh adapter — which is the very thing these tests exist to rule out. Each now fails when its corresponding fix is reverted.
